File: server/src/main/java/com/bomberman/SocketServer.java
package com.bomberman;

import com.bomberman.entities.Movement;
import com.bomberman.entities.mapobject.Bomb;
import com.bomberman.game.Game;
import com.corundumstudio.socketio.Configuration;
import com.corundumstudio.socketio.SocketIOClient;
import com.corundumstudio.socketio.SocketIOServer;
import java.util.concurrent.ScheduledThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

class SocketServer {
  private final Game game;
  private SocketIOServer server;

  SocketServer(final Game game) {
    this.game = game;

    initServer();
  }

  private static String getPlayerId(final SocketIOClient client) {
    return client.getSessionId().toString();
  }

  private void initServer() {
    final var config = new Configuration();
    config.setPort(5050);

    server = new SocketIOServer(config);

    addEventListeners();

    server.start();
    System.out.println("socket server has started");
  }

  private void addEventListeners() {
    server.addEventListener(
        "connectNewPlayer",
        String.class,
        (client, data, ackRequest) -> {
          final var playerId = getPlayerId(client);
          System.out.println("New player connected: " + playerId);
          game.addPlayer(playerId);
          client.sendEvent("connectNewPlayerSuccess", playerId);
          server.getBroadcastOperations().sendEvent("getGameState", game.getGameState());
        });

    server.addEventListener(
        "disconnectPlayer",
        String.class,
        (client, data, ackRequest) -> {
          final var playerId = getPlayerId(client);
          game.removePlayer(playerId);
          client.sendEvent("disconnectPlayerSuccess");
          server.getBroadcastOperations().sendEvent("getGameState", game.getGameState());
        });

    server.addEventListener(
        "movePlayer",
        Movement.class,
        (client, movement, ackRequest) -> {
          final var playerId = getPlayerId(client);
          game.movePlayer(playerId, movement);
          server.getBroadcastOperations().sendEvent("getGameState", game.getGameState());
        });

    server.addEventListener(
        "placeBomb",
        String.class,
        (client, data, ackRequest) -> {
          final var playerId = getPlayerId(client);
          final var bomb = game.placeBomb(playerId);
          server.getBroadcastOperations().sendEvent("getGameState", game.getGameState());
          bomb.ifPresent(this::setBombExplosion);
        });

    // DEBUG
    server.addEventListener(
        "respawnPlayer",
        String.class,
        (client, data, ackRequest) -> {
          final var playerId = getPlayerId(client);
          game.respawnPlayer(playerId);
          server.getBroadcastOperations().sendEvent("getGameState", game.getGameState());
        });
    server.addEventListener(
        "increaseBombRange",
        String.class,
        (client, data, ackRequest) -> {
          final var playerId = getPlayerId(client);

          game.increaseBombRange(playerId);
          server.getBroadcastOperations().sendEvent("getGameState", game.getGameState());
        });
    server.addEventListener(
        "decreaseBombRange",
        String.class,
        (client, data, ackRequest) -> {
          final var playerId = getPlayerId(client);
          game.decreaseBombRange(playerId);
          server.getBroadcastOperations().sendEvent("getGameState", game.getGameState());
        });
    server.addEventListener(
        "increasePlayerSpeed",
        String.class,
        (client, data, ackRequest) -> {
          final var playerId = getPlayerId(client);
          game.increasePlayerSpeed(playerId);
          server.getBroadcastOperations().sendEvent("getGameState", game.getGameState());
        });
    server.addEventListener(
        "decreasePlayerSpeed",
        String.class,
        (client, data, ackRequest) -> {
          final var playerId = getPlayerId(client);
          game.decreasePlayerSpeed(playerId);
          server.getBroadcastOperations().sendEvent("getGameState", game.getGameState());
        });
    server.addEventListener(
        "increaseMaxBombCount",
        String.class,
        (client, data, ackRequest) -> {
          final var playerId = getPlayerId(client);
          game.increaseMaxBombCount(playerId);
          server.getBroadcastOperations().sendEvent("getGameState", game.getGameState());
        });
  }

  private void setBombExplosion(final Bomb bomb) {
    final ScheduledThreadPoolExecutor executor = new ScheduledThreadPoolExecutor(1);
    executor.schedule(
        () -> {
          game.explodeBomb(bomb, bomb.getBombId(), bomb.getPlayerId());
          game.checkKilled(bomb);
          removeBombExplosion(bomb);
          server.getBroadcastOperations().sendEvent("getGameState", game.getGameState());
        },
        bomb.getExplosionTime(),
        TimeUnit.SECONDS);
  }

  private void removeBombExplosion(final Bomb bomb) {
    final ScheduledThreadPoolExecutor executor = new ScheduledThreadPoolExecutor(1);
    executor.schedule(
        () -> {
          game.removeExplosions(bomb);
          server.getBroadcastOperations().sendEvent("getGameState", game.getGameState());
        },
        1,
        TimeUnit.SECONDS);
  }
}

File: server/src/main/kotlin/com/bomberman/SocketServer.kt
package com.bomberman

import com.bomberman.entity.Direction
import com.bomberman.entity.mapobject.Bomb
import com.bomberman.game.Game
import com.corundumstudio.socketio.Configuration
import com.corundumstudio.socketio.SocketIOClient
import com.corundumstudio.socketio.SocketIOServer
import java.util.concurrent.ScheduledThreadPoolExecutor
import java.util.concurrent.TimeUnit

class SocketServer(
    private val game: Game
) {
  private val server: SocketIOServer

  init {
    val config = Configuration()
    config.port = 5050

    server = SocketIOServer(config)

    addEventListeners()

    server.start()
    println("Socket server has started")
  }

  private fun getPlayerId(client: SocketIOClient) =
      client.sessionId.toString()

  private fun addEventListeners() {
    server.addEventListener(
        "connectNewPlayer",
        String::class.java
    ) { client, _, _ ->
      val playerId = getPlayerId(client)
      if (game.playerExists(playerId)) {
        client.sendEvent("connectNewPlayerSuccess", playerId)
        server.broadcastOperations.sendEvent("getGameState", game.gameState)
        return@addEventListener
      }
      println("New player connected: $playerId")
      game.addPlayer(playerId)
      client.sendEvent("connectNewPlayerSuccess", playerId)
      server.broadcastOperations.sendEvent("getGameState", game.gameState)
    }

    server.addEventListener(
        "disconnectPlayer",
        String::class.java
    ) { client, _, _ ->
      val playerId = getPlayerId(client)
      game.removePlayer(playerId)
      client.sendEvent("disconnectPlayerSuccess")
      println("Player disconnected: $playerId")
      server.broadcastOperations.sendEvent("getGameState", game.gameState)
    }

    server.addEventListener(
        "movePlayer",
        Direction::class.java
    ) { client, direction, _ ->
      val playerId = getPlayerId(client)
      game.movePlayer(playerId, direction)
      server.broadcastOperations.sendEvent("getGameState", game.gameState)
    }

    server.addEventListener(
        "placeBomb",
        String::class.java
    ) { client, _, _ ->
      val playerId = getPlayerId(client)
      val bomb = game.placeBomb(playerId)
      server.broadcastOperations.sendEvent("getGameState", game.gameState)
      bomb?.setBombExplosion()
    }

    // DEBUG
    server.addEventListener(
        "respawnPlayer",
        String::class.java
    ) { client, _, _ ->
      val playerId = getPlayerId(client)
      game.respawnPlayer(playerId)
      server.broadcastOperations.sendEvent("getGameState", game.gameState)
    }
    server.addEventListener(
        "increaseBombRange",
        String::class.java
    ) { client, _, _ ->
      val playerId = getPlayerId(client)
      game.increaseBombRange(playerId)
      server.broadcastOperations.sendEvent("getGameState", game.gameState)
    }
    server.addEventListener(
        "decreaseBombRange",
        String::class.java
    ) { client, _, _ ->
      val playerId = getPlayerId(client)
      game.decreaseBombRange(playerId)
      server.broadcastOperations.sendEvent("getGameState", game.gameState)
    }
    server.addEventListener(
        "increasePlayerSpeed",
        String::class.java
    ) { client, _, _ ->
      val playerId = getPlayerId(client)
      game.increasePlayerSpeed(playerId)
      server.broadcastOperations.sendEvent("getGameState", game.gameState)
    }
    server.addEventListener(
        "decreasePlayerSpeed",
        String::class.java
    ) { client, _, _ ->
      val playerId = getPlayerId(client)
      game.decreasePlayerSpeed(playerId)
      server.broadcastOperations.sendEvent("getGameState", game.gameState)
    }
    server.addEventListener(
        "increaseMaxBombCount",
        String::class.java
    ) { client, _, _ ->
      val playerId = getPlayerId(client)
      game.increaseMaxBombCount(playerId)
      server.broadcastOperations.sendEvent("getGameState", game.gameState)
    }

  }

  private fun Bomb.setBombExplosion() {
    val executor = ScheduledThreadPoolExecutor(1)
    executor.schedule(
        {
          game.explodeBomb(this, this.bombId, this.playerId)
          game.checkKilled(this)
          this.removeBombExplosion()
          server.broadcastOperations.sendEvent("getGameState", game.gameState)
        },
        this.explosionTime.toLong(),
        TimeUnit.SECONDS)

  }

  private fun Bomb.removeBombExplosion() {
    val executor = ScheduledThreadPoolExecutor(1)
    executor.schedule(
        {
          game.removeExplosions(this)
          server.broadcastOperations.sendEvent("getGameState", game.gameState)
        },
        1,
        TimeUnit.SECONDS)
  }
}
